src/Document/CategoryManager.php

```diff
@@ -157,36 +108,87 @@ public function getPager(array $criteria, int $page, int $limit = 10, array $sor
         return $pager;
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function getRootCategoryWithChildren(CategoryInterface $category): void
+    public function getRootCategoryWithChildren(CategoryInterface $category)
+    {
+        if (null === $category->getContext()) {
+            throw new \InvalidArgumentException('Context cannot be null.');
+        }
+        if (null !== $category->getParent()) {
+            throw new \InvalidArgumentException('Method can be called only for root categories.');
+        }
+
+        $context = $category->getContext();
+
+        $this->loadCategories($context);
+
+        foreach ($this->categories[$context->getId()] as $contextRootCategory) {
+            if ($category->getId() === $contextRootCategory->getId()) {
+                return $contextRootCategory;
+            }
+        }
+
+        throw new \InvalidArgumentException(sprintf('Category "%s" does not exist.', $category->getId()));
+    }
+
+    public function getRootCategoriesForContext(?ContextInterface $context = null)
     {
-        throw new \RuntimeException('Not Implemented yet');
+        $context = $this->getContext($context);
+
+        $this->loadCategories($context);
+
+        return $this->categories[$context->getId()];
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function getRootCategoriesForContext(?ContextInterface $context = null): void
+    public function getAllRootCategories($loadChildren = true)
     {
-        throw new \RuntimeException('Not Implemented yet');
+        $rootCategories = $this->getDocumentManager()
+            ->getRepository($this->class)
+            ->createQueryBuilder('c')
+            ->where('c.parent IS NULL')
+            ->getQuery()
+            ->getResult();
+
+        $categories = [];
+
+        foreach ($rootCategories as $category) {
+            if (null === $category->getContext()) {
+                throw new \LogicException('Context cannot be null.');
```

Comment:
IMO, this message is misleading since there is no way to know which category is missing the context.

Reply:
Done